qcore/asserts.py

```diff
@@ -138,11 +139,20 @@ def _dict_path_string(path):
     return "->".join(map(ascii, path))
 
 
+def _defaultdict_to_dict(value):
+    if isinstance(value, defaultdict):
+        value = {k: _defaultdict_to_dict(v) for k, v in value.items()}
+    return value
+
+
 def assert_dict_eq(expected, actual, number_tolerance=None, dict_path=[]):
     """Asserts that two dictionaries are equal, producing a custom message if they are not."""
     assert_is_instance(expected, dict)
     assert_is_instance(actual, dict)
 
+    expected = _defaultdict_to_dict(expected)
+    actual = _defaultdict_to_dict(actual)
+
```

Comment:
This is an odd special case. What if instead we skip the two assert_is_instance() calls on line 159 below if `actual[k] == expected[k]`?

Reply:
Okay, generalized the logic to all dict-subclasses, while keeping the performance for dict values.

Although, the solution still would have a problem if there's dict-subclass nested under a proper dict.

Wdyt?

Reply:
> This is an odd special case. What if instead we skip the two assert_is_instance() calls on line 159 below if actual[k] == expected[k]?

A problem with this approach would be that it would be hard to see what the mismatch is when that condition fails.